diff/fd/hessian.go

```diff
@@ -11,28 +11,26 @@ import (
 	"gonum.org/v1/gonum/mat"
 )
 
-// Hessian approximates the Hessian matrix of the multivariate function f
-// at the location x. That is
+// Hessian approximates the Hessian matrix of the multivariate function f at
+// the location x. That is
 //  H_{i,j} = ∂^2 f(x)/∂x_i ∂x_j
-// If dst is not nil, the resulting H will be stored in-place into dst
-// and returned, otherwise a new matrix will be allocated first. Finite difference
-// formula and other options are specified by settings. If settings is nil,
-// the Hessian will be estimated using the Forward formula and a default step size.
+// The resulting H will be stored in dst. Finite difference formula and other
+// options are specified by settings. If settings is nil, the Hessian will be
+// estimated using the Forward formula and a default step size.
 //
-// Hessian panics if the size of dst and x is not equal, or if the derivative
-// order of the formula is not 1.
-func Hessian(dst *mat.SymDense, f func(x []float64) float64, x []float64, settings *Settings) *mat.SymDense {
+// If the dst matrix is zero-sized it will be resized to the correct dimensions,
+// otherwise dst must match the dimension of the receiver or CovarianceMatrix
+// will panic. Hessian will panic if the derivative order of the formula is not 1.
+func Hessian(dst *mat.SymDense, f func(x []float64) float64, x []float64, settings *Settings) {
 	n := len(x)
-	if dst == nil {
-		dst = mat.NewSymDense(n, nil)
-	} else {
-		if n2 := dst.Symmetric(); n2 != n {
-			panic("hessian: dst size mismatch")
-		}
-		for i := 0; i < n; i++ {
-			for j := i; j < n; j++ {
-				dst.SetSym(i, j, 0)
-			}
+	if dst.IsZero() {
+		*dst = *(dst.GrowSym(n).(*mat.SymDense))
+	} else if dst.Symmetric() != n {
+		panic("hessian: dst size mismatch")
+	}
+	for i := 0; i < n; i++ {
```

Comment:
While here, just dst.Zero()?

Reply:
I forgot we had that. I was wishing for one, but didn't bother looking.